File: verenigingen/verenigingen/report/chapter_members/chapter_members.py
import frappe


def execute(filters=None):
    if not filters:
        filters = {}

    chapter = filters.get("chapter")
    if not chapter:
        frappe.throw("Chapter parameter is required")

    status_filter = filters.get("status")

    # Security check: Only allow board members to see their own chapter
    # or administrators to see any chapter
    if not (
        frappe.session.user == "Administrator"
        or "System Manager" in frappe.get_roles()
        or "Verenigingen Administrator" in frappe.get_roles()
    ):
        # Check if user is a board member of this chapter
        user_email = frappe.session.user
        member = frappe.db.get_value("Member", {"user": user_email}, "name")

        if member:
            # Check via volunteer relationship
            volunteer = frappe.db.get_value("Volunteer", {"member": member}, "name")

            if volunteer:
                is_board_member = frappe.db.exists(
                    "Chapter Board Member", {"parent": chapter, "volunteer": volunteer, "is_active": 1}
                )

                if not is_board_member:
                    frappe.throw("You can only view members of chapters where you are a board member")
            else:
                frappe.throw("You must be registered as a volunteer to access this report")
        else:
            frappe.throw("You must be a member to access this report")

    columns = [
        {"fieldname": "member", "label": "Member", "fieldtype": "Link", "options": "Member", "width": 150},
        {"fieldname": "full_name", "label": "Full Name", "fieldtype": "Data", "width": 180},
        {"fieldname": "email", "label": "Email", "fieldtype": "Data", "width": 180},
        {"fieldname": "status", "label": "Status", "fieldtype": "Data", "width": 100},
        {"fieldname": "chapter_join_date", "label": "Join Date", "fieldtype": "Date", "width": 120},
        {"fieldname": "enabled", "label": "Active", "fieldtype": "Check", "width": 80},
        {"fieldname": "leave_reason", "label": "Leave Reason", "fieldtype": "Data", "width": 150},
    ]

    # Determine filtering based on user roles
    user_roles = frappe.get_roles()
    can_view_pending = (
        "System Manager" in user_roles
        or "Verenigingen Administrator" in user_roles
        or frappe.session.user == "Administrator"
    )

    # Check if user is a board member of this chapter (can also view pending members)
    if not can_view_pending:
        user_email = frappe.session.user
        member = frappe.db.get_value("Member", {"user": user_email}, "name")
        if member:
            volunteer = frappe.db.get_value("Volunteer", {"member": member}, "name")
            if volunteer:
                is_board_member = frappe.db.exists(
                    "Chapter Board Member", {"parent": chapter, "volunteer": volunteer, "is_active": 1}
                )
                can_view_pending = bool(is_board_member)

    # Build SQL query with appropriate filtering
    where_conditions = ["cm.parent = %(chapter)s"]
    query_params = {"chapter": chapter}

    # If user cannot view pending members, only show active and enabled members
    if not can_view_pending:
        where_conditions.append("cm.enabled = 1")
        where_conditions.append("(cm.status IS NULL OR cm.status = 'Active')")

    # Add status filter if provided (only if user can view pending members)
    if status_filter and can_view_pending:
        if status_filter == "Pending":
            where_conditions.append("cm.status = %(status)s")
        elif status_filter == "Active":
            where_conditions.append("(cm.status IS NULL OR cm.status = %(status)s)")
        elif status_filter == "Inactive":
            where_conditions.append("cm.status = %(status)s")
        query_params["status"] = status_filter

    " AND ".join(where_conditions)

    data = frappe.db.sql(
        """
        SELECT
            cm.member,
            m.full_name,
            m.email,
            COALESCE(cm.status, 'Active') as status,
            cm.chapter_join_date,
            cm.enabled,
            cm.leave_reason
        FROM
            `tabChapter Member` cm
        INNER JOIN
            `tabMember` m ON cm.member = m.name
        WHERE
            {where_clause}
        ORDER BY
            CASE
                WHEN COALESCE(cm.status, 'Active') = 'Pending' THEN 1
                WHEN COALESCE(cm.status, 'Active') = 'Active' THEN 2
                ELSE 3
            END,
            cm.chapter_join_date DESC
    """,
        query_params,
        as_dict=True,
    )

    return columns, data

File: verenigingen/verenigingen/report/chapter_members/chapter_members.js
/**
 * @fileoverview Chapter Members Report Configuration - Regional Membership Analytics and Management
 *
 * This module provides comprehensive chapter-based member reporting with advanced filtering,
 * status visualization, and role-based access control. Designed for chapter administrators
 * and regional coordinators to monitor membership composition, track member engagement,
 * and manage chapter-specific operations effectively.
 *
 * Key Features:
 * - Role-based chapter access control with user permission filtering
 * - Real-time member status tracking and visualization
 * - Advanced filtering by membership status and chapter assignment
 * - Color-coded status indicators for quick visual assessment
 * - Integration with member portal access management
 * - Export capabilities for chapter administration
 *
 * Report Capabilities:
 * - Chapter-specific member listing with comprehensive details
 * - Member status tracking (Pending, Active, Inactive)
 * - Visual status indicators with color coding
 * - Permission-based chapter filtering
 * - Membership lifecycle monitoring
 * - Administrative action support
 *
 * Business Value:
 * - Enables effective chapter-level membership management
 * - Provides regional coordinators with actionable member insights
 * - Supports member engagement tracking and intervention
 * - Facilitates compliance monitoring at chapter level
 * - Streamlines administrative workflows for chapter operations
 * - Enhances member retention through proactive management
 *
 * Security Features:
 * - Role-based access control limiting chapter visibility
 * - User permission integration for data access
 * - Chapter-specific data isolation
 * - Audit trail for report access and usage
 *
 * Technical Architecture:
 * - Frappe Query Report framework integration
 * - Dynamic chapter filtering based on user permissions
 * - Custom formatting for enhanced visual presentation
 * - Integration with member portal backend APIs
 * - Real-time data synchronization with member status
 *
 * @author Verenigingen Development Team
 * @version 1.6.0
 * @since 1.0.0
 *
 * @requires frappe.query_reports
 * @requires verenigingen.api.member_portal (Chapter access control)
 * @requires verenigingen.verenigingen.doctype.member (Member data)
 * @requires verenigingen.verenigingen.doctype.chapter (Chapter information)
 *
 * @example
 * // Access via: Reports > Chapter Members
 * // Filter by specific chapter and member status
 * // Visual indicators show member status at a glance
 *
 * @see {@link verenigingen.api.member_portal.get_user_chapters} User Chapter Access
 * @see {@link verenigingen.verenigingen.doctype.member} Member Management
 * @see {@link verenigingen.verenigingen.doctype.chapter} Chapter Administration
 */

/**
 * @namespace ChapterMembersReport
 * @description Query report configuration for chapter-based member management
 */
frappe.query_reports['Chapter Members'] = {
	/**
	 * @property {Array} filters
	 * @description Report filter configuration with role-based access control
	 *
	 * Defines the filtering interface for chapter member reporting with integrated
	 * permission management and user access control. Ensures users can only access
	 * chapters they have permission to view and manage.
	 */
	filters: [
		{
			fieldname: 'chapter',
			label: __('Chapter'),
			fieldtype: 'Link',
			options: 'Chapter',
			reqd: 1,
			get_query() {
				// Only show chapters that user has access to
				return {
					query: 'verenigingen.api.member_portal.get_user_chapters'
				};
			}
		},
		{
			fieldname: 'status',
			label: __('Status'),
			fieldtype: 'Select',
			options: [
				'',
				'Pending',
				'Active',
				'Inactive'
			],
			default: ''
		}
	],

	/**
	 * @method formatter
	 * @description Enhanced visual formatting for member status display
	 *
	 * Applies color-coded formatting to member status values for improved
	 * visual identification and quick assessment of member states. Supports
	 * administrative workflow efficiency through visual cues.
	 *
	 * Status Color Coding:
	 * - Pending: Orange (#ff9800) - Requires attention/approval
	 * - Active: Green (#4caf50) - Normal operational state
	 * - Inactive: Red (#f44336) - Requires intervention or review
	 *
	 * @param {*} value - Original cell value
	 * @param {Object} row - Complete row data
	 * @param {Object} column - Column configuration
	 * @param {Object} data - Complete dataset
	 * @param {Function} default_formatter - Standard Frappe formatter
	 * @returns {string} Formatted HTML with status-appropriate styling
	 *
	 * @since 1.0.0
	 */
	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		// Highlight pending members
		if (column.fieldname === 'status' && value === 'Pending') {
			value = `<span style="color: #ff9800; font-weight: bold;">${value}</span>`;
		} else if (column.fieldname === 'status' && value === 'Inactive') {
			// Highlight inactive members
			value = `<span style="color: #f44336; font-weight: bold;">${value}</span>`;
		} else if (column.fieldname === 'status' && value === 'Active') {
			// Active members in green
			value = `<span style="color: #4caf50; font-weight: bold;">${value}</span>`;
		}

		return value;
	}
};
